### PR TITLE
Show Java source level in the status bar

### PR DESCRIPTION
**Pull request by jdneo**
requires: https://github.com/eclipse/eclipse.jdt.ls/pull/1447

Kapture 2020-05-19 at 14 38 57

1. The status bar item will only show up when current active editor is `.java` file
2. The status bar item is used to show the project's source level
3. In the future the status bar item might be used as an entry to expose more commands

requires https://github.com/eclipse/eclipse.jdt.ls/pull/1447

Signed-off-by: Sheng Chen <sheche@microsoft.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -453,7 +453,7 @@
         },
         "java.configuration.runtimes": {
           "type": "array",
-          "label": "Map Java Execution Environments to local JDKs",
+          "description": "Map Java Execution Environments to local JDKs",
```
Comment (jdneo) [inline package.json:456]: @fbricon I guess the `label` is a typo?

Comment (fbricon) [reply package.json:456]: Yes

File: src/runtimeStatusBarProvider.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		}
+
+		if (belongingProjectPath) {
+			this.fileProjectMapping.set(uri.fsPath, belongingProjectPath);
+		}
+
+		return belongingProjectPath;
+	}
+
+	private async getProjectInfo(projectPath: string): Promise<IProjectInfo> {
```
Comment (fbricon) [inline src/runtimeStatusBarProvider.ts:91]: const projectInfo = this.javaProjects.get(projectPath);
if (projectInfo) {
   return projectInfo;
}

File: src/runtimeStatusBarProvider.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	public dispose(): void {
+		this.statusBarItem.dispose();
+		for (const disposable of this.disposables) {
+			disposable.dispose();
+		}
+	}
+
+	private findBelongingProject(uri: Uri): string {
```
Comment (fbricon) [inline src/runtimeStatusBarProvider.ts:65]: findOwnerProject

File: package.json [hunk 4]
```diff
@@ -476,29 +476,28 @@
...
               },
               "path": {
                 "type": "string",
-                "label": "JDK path.\nOn Windows, backslashes must be escaped, i.e.\n\"path\":\"C:\\\\Program Files\\\\Java\\\\jdk1.8.0_161\""
+                "description": "JDK path.\nOn Windows, backslashes must be escaped, i.e.\n\"path\":\"C:\\\\Program Files\\\\Java\\\\jdk1.8.0_161\""
               },
               "sources": {
                 "type": "string",
-                "label": "JDK sources."
+                "description": "JDK sources."
```
Comment (fbricon) [inline package.json:487]: JDK sources path

File: package.json [hunk 5]
```diff
@@ -476,29 +476,28 @@
...
               },
               "javadoc": {
                 "type": "string",
-                "label": "JDK javadoc."
+                "description": "JDK javadoc."
```
Comment (fbricon) [inline package.json:491]: JDK javadoc path

File: src/runtimeStatusBarProvider.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+			return;
+		}
+
+		for (const uri of projectUriStrings) {
+			this.javaProjects.set(Uri.parse(uri).fsPath, undefined);
+		}
+
+		this.statusBarItem.command = {
+			title: "Configure Java Runtime",
+			command: "workbench.action.openSettings",
```
Comment (fbricon) [inline src/runtimeStatusBarProvider.ts:36]: I think it should open settings.json file directly, not the UI, as you need to manually switch to the json anyways

Comment (fbricon) [reply src/runtimeStatusBarProvider.ts:36]: scratch that. Starting from the UI will be easier if java.configuration.runtimes is not set

File: src/extension.ts [hunk 7]
```diff
@@ -286,6 +287,7 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 						switch (report.type) {
 							case 'ServiceReady':
 								syntaxClient.stop();
+								runtimeStatusBarProvider.initialize(onDidClasspathUpdate, context.storagePath);
```
Comment (fbricon) [inline src/extension.ts:290]: when opening a standalone java file directly (with no opened folder), this leads to:
```
rejected promise not handled within 1 second: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
extensionHostProcess.js:957
stack trace: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
	at validateString (internal/validators.js:112:11)
	at Object.join (path.js:1040:7)
	at RuntimeStatusBarProvider.<anonymous> (/Users/fbricon/Dev/projects/vscode-java/dist/extension.js:72775:55)
	at Generator.next (<anonymous>)
	at /Users/fbricon/Dev/projects/vscode-java/dist/extension.js:72759:71
```

**Pull request merged**